checkpatch_hook/__init__.py

```diff
@@ -0,0 +1,252 @@
+import sys
+from collections import defaultdict
+from datetime import datetime
+from pathlib import Path
+from subprocess import PIPE, check_output, run
+from typing import DefaultDict
+
+import yaml
+
+# template used to add a patch-style header to a diff
+PATCH_TEMPLATE = """\
+From: A Non <a.non@bluwireless.com>
+Date: {current_datetime}
+Subject: [PATCH] {commit_msg}
+---
+{diff}
+--
+"""
+
+
+def _parse_args() -> dict:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--ci",
+        action="store_true",
+        help="Indicates that the (manual) hook is being run in CI",
+    )
+    parser.add_argument(
+        "files",
+        nargs="*",
+        type=Path,
+        default=[],
+        help="Files to check",
+    )
+    default_config_file = importlib.resources.files("checkpatch_hook.data").joinpath(
+        "checkpatch.yaml"
+    )
+    parser.add_argument(
+        "--config-file",
+        type=Path,
+        default=default_config_file,
+        help="Path to config file",
+    )
+    return vars(parser.parse_args())
+
+
+class ConfigFile:
+    DEFAULT_DIR_KEY = "default"
+    mandatory_keys = ["DIR_CONFIGS"]
+    magic_error_keys = {"errors_enabled": "ERRORS_COMMON", "errors_ignored": "IGNORES_COMMON"}
+    optional_keys = ["IGNORED_FILES"]
+    # TODO: Use schema based validation e.g., cfgv
+
+    def __init__(self, config_file: Path):
+        self.config_file = config_file
+
+    def load_config(self) -> dict:
+        try:
+            with self.config_file.open("r") as file:
+                config = yaml.safe_load(file)
+            if config is None or not isinstance(config, dict) or not config:
+                raise RuntimeError("Config file is empty")
+            for key in self.mandatory_keys:
+                if key not in config:
+                    raise RuntimeError(f"Missing mandatory key {key} in {self.config_file}")
+                if not isinstance(config[key], dict) or not config[key]:
+                    raise RuntimeError(f"Invalid type for key {key} in {self.config_file}")
+            # Check for unknown keys
+            for key in config.keys():
+                if key not in self.mandatory_keys + self.optional_keys + list(
+                    self.magic_error_keys.values()
+                ):
+                    raise RuntimeError(f"Unknown key {key} in {self.config_file}")
+            if not isinstance(config.get("IGNORED_FILES", []), list):
+                raise RuntimeError(f"Invalid type for key IGNORED_FILES in {self.config_file}")
+            return config
+        except FileNotFoundError:
+            print(f"Config file not found: {self.config_file}", file=sys.stderr)
+            sys.exit(1)
+        except yaml.YAMLError as e:
+            print(f"Error loading YAML config file: {e}", file=sys.stderr)
+            sys.exit(1)
+        except RuntimeError as e:
+            print(f"Error loading YAML config file: {e}", file=sys.stderr)
```

Comment:
At the moment the strings you are providing above when you `raise RuntimeError(...)` are being lost. I think what I was asking for earlier would probably require replacing the use of `RuntimeError` with a custom exception type that lets you retrieve the message and print that:

```python
class _ConfigException(BaseException):
    def __init__(..., msg: str):
        super().__init__(...)
        self.msg = msg
```

then you just print `e.msg` in that particular `except` block.

BTW, is printing to `sys.stderr` necessary? Maybe see if there's a convention for where some widely-used hooks handle that kinda thing.

Reply:
The strings are not lost, as I printing `{e}`, see

```
$ checkpatch --config-file test.yaml
Error loading YAML config file: Missing mandatory key DIR_CONFIGS in test.yaml
```

> BTW, is printing to sys.stderr necessary? Maybe see if there's a convention for where some widely-used hooks handle that kinda thing.

Yes, errors are being printed to stderr in most hooks.

Reply:
Ah I didn't realise that printing an exception behaved like that - well that looks good to me then if you're happy too.